File: src/javascript/datastructures/singlyLinkedList/singlyLinkedList.js
/**
 * Node class for singly linked list.
 */
class Node {
  /**
   * @param {*} value
   */
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

/**
 * Singly Linked List implementation.
 */
class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /**
   * Adds a new node with the given value at the end of the list.
   * @param {*} value
   * @returns {LinkedList}
   */
  push(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length++;
    return this;
  }

  /**
   * Removes the last node from the list and returns it.
   * @returns {Node|undefined}
   */
  pop() {
    if (!this.head) return undefined;
    let current = this.head;
    let newTail = this.head;
    while (current.next) {
      newTail = current;
      current = current.next;
    }
    this.tail = newTail;
    this.tail.next = null;
    this.length--;
    if (this.length === 0) {
      this.head = null;
      this.tail = null;
    }
    return current;
  }

  /**
   * Adds a new node with the given value at the start of the list.
   * @param {*} value
   * @returns {LinkedList}
   */
  unshift(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length++;
    return this;
  }

  /**
   * Removes the first node from the list and returns it.
   * @returns {Node|undefined}
   */
  shift() {
    if (!this.head) return undefined;
    let currentHead = this.head;
    this.head = this.head.next;
    currentHead.next = null;
    this.length--;
    if (this.length === 0) {
      this.tail = null;
    }
    return currentHead;
  }

  /**
   * Returns the node at the given index.
   * @param {number} index
   * @returns {Node|undefined}
   */
  get(index) {
    if (index < 0 || index >= this.length) return undefined;
    let current = this.head;
    for (let i = 0; i < index; i++) {
      current = current.next;
    }
    return current;
  }

  /**
   * Updates the value of the node at the given index.
   * @param {number} index
   * @param {*} value
   * @returns {boolean}
   */
  set(index, value) {
    const node = this.get(index);
    if (node) {
      node.value = value;
      return true;
    }
    return false;
  }

  /**
   * Inserts a new node with the given value at the specified index.
   * @param {number} index
   * @param {*} value
   * @returns {boolean}
   */
  insert(index, value) {
    if (index < 0 || index > this.length) return false;
    if (index === 0) return !!this.unshift(value);
    if (index === this.length) return !!this.push(value);

    const newNode = new Node(value);
    const prevNode = this.get(index - 1);
    newNode.next = prevNode.next;
    prevNode.next = newNode;
    this.length++;
    return true;
  }

  /**
   * Removes the node at the specified index and returns it.
   * @param {number} index
   * @returns {Node|undefined}
   */
  remove(index) {
    if (index < 0 || index >= this.length) return undefined;
    if (index === 0) return this.shift();
    if (index === this.length - 1) return this.pop();

    const prevNode = this.get(index - 1);
    const removedNode = prevNode.next;
    prevNode.next = removedNode.next;
    removedNode.next = null;
    this.length--;
    return removedNode;
  }

  /**
   * Reverses the linked list in place.
   * @returns {LinkedList}
   */
  reverse() {
    if(this.length === 0) return undefined;

    let temp = this.head;
    this.head = this.tail;
    this.tail = temp;
    let next = temp.next;
    let prev = null;
    for (let i = 0; i < this.length; i++) {
      next = temp.next;
      temp.next = prev;
      prev = temp;
      temp = next;
    }
    return this;
  }

  /**
   * Converts the linked list to an array.
   * @returns {Array}
   */
  toArray() {
    const arr = [];
    let current = this.head;
    while (current) {
      arr.push(current.value);
      current = current.next;
    }
    return arr;
  }

  /**
   * Prints the linked list as a string.
   * @returns {string}
   */
  printList() {
    return this.toArray().join(" -> ");
  }
}

// Usage example
const list = new LinkedList();
list.push(1).push(2).push(3);
console.log("List:", list.printList());

console.log("Pop:", list.pop());
console.log("Pop:", list.pop());
console.log("After pop:", list.printList());

list.unshift(2);
list.unshift(3);
console.log("After unshift:", list.printList());

console.log(list.shift());
console.log("After shift:", list.printList());

console.log(list.get(1));
console.log(list.get(0));

console.log(list.set(0, 1));
console.log(list.set(1, 3));
console.log("After set:", list.printList());

list.insert(1, 2);
list.insert(3, 4);
console.log("After insert:", list.printList());

console.log(list.remove(3));
console.log(list.remove(2));
console.log("After Remove:", list.printList());

list.reverse();
console.log("After reverse:", list.printList());
list.reverse();
console.log("After reverse:", list.printList());

File: src/python/datastructures/singlyLinkedList/singly_linked_list.py
"""
Singly Linked List implementation in Python.
"""


class Node:
    """Node class for singly linked list."""
    
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    """Singly Linked List implementation."""
    
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def push(self, value):
        """Adds a new node with the given value at the end of the list."""
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1
        return self

    def pop(self):
        """Removes the last node from the list and returns it."""
        if not self.head:
            return None
        current = self.head
        new_tail = self.head
        while current.next:
            new_tail = current
            current = current.next
        self.tail = new_tail
        self.tail.next = None
        self.length -= 1
        if self.length == 0:
            self.head = None
            self.tail = None
        return current

    def unshift(self, value):
        """Adds a new node with the given value at the start of the list."""
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        self.length += 1
        return self

    def shift(self):
        """Removes the first node from the list and returns it."""
        if not self.head:
            return None
        current_head = self.head
        self.head = self.head.next
        current_head.next = None
        self.length -= 1
        if self.length == 0:
            self.tail = None
        return current_head

    def get(self, index):
        """Returns the node at the given index."""
        if index < 0 or index >= self.length:
            return None
        current = self.head
        for _ in range(index):
            current = current.next
        return current

    def set(self, index, value):
        """Updates the value of the node at the given index."""
        node = self.get(index)
        if node:
            node.value = value
            return True
        return False

    def insert(self, index, value):
        """Inserts a new node with the given value at the specified index."""
        if index < 0 or index > self.length:
            return False
        if index == 0:
            return bool(self.unshift(value))
        if index == self.length:
            return bool(self.push(value))

        new_node = Node(value)
        prev_node = self.get(index - 1)
        new_node.next = prev_node.next
        prev_node.next = new_node
        self.length += 1
        return True

    def remove(self, index):
        """Removes the node at the specified index and returns it."""
        if index < 0 or index >= self.length:
            return None
        if index == 0:
            return self.shift()
        if index == self.length - 1:
            return self.pop()

        prev_node = self.get(index - 1)
        removed_node = prev_node.next
        prev_node.next = removed_node.next
        removed_node.next = None
        self.length -= 1
        return removed_node

    def reverse(self):
        """Reverses the linked list in place."""
        current = self.head
        self.head, self.tail = self.tail, self.head
        prev = None
        for _ in range(self.length):
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        return self

    def print_list(self):
        """Prints all values in the list."""
        values = []
        current = self.head
        while current:
            values.append(str(current.value))
            current = current.next
        print(" -> ".join(values))


# Usage example
if __name__ == "__main__":
    ll = LinkedList()
    ll.push(1)
    ll.push(2)
    ll.push(3)
    ll.push(4)

    print("Original list:")
    ll.print_list()

    print("\nAfter reversing:")
    ll.reverse()
    ll.print_list()

    print(f"\nLength: {ll.length}")
    print(f"Head value: {ll.head.value}")
    print(f"Tail value: {ll.tail.value}")
